<a href="https://colab.research.google.com/github/eshan14git/football-qa-nlp/blob/eshan-dev/notebooks/01_data_exploration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Football Question Answering Assistant

## Phase 1 - Data Exploration (EDA)

### Objectives
- Explore all project datasets
- Understand dataset relationships
- Identify missing values and duplicates
- Document dataset structure
- Prepare for preprocessing

**Prepared by:** Eshan Menuka

**Module:** CCS3356 – Natural Language Processing

In [6]:
import pandas as pd
import numpy as np

print("✅ Libraries imported successfully!")

✅ Libraries imported successfully!


In [7]:
!git clone -b eshan-dev https://github.com/eshan14git/football-qa-nlp.git

fatal: destination path 'football-qa-nlp' already exists and is not an empty directory.


In [8]:
import os

print(os.listdir("football-qa-nlp/data"))

['.gitkeep']


In [9]:
!rm -rf football-qa-nlp

In [10]:
!git clone -b eshan-dev https://github.com/eshan14git/football-qa-nlp.git

Cloning into 'football-qa-nlp'...
remote: Enumerating objects: 140, done.
remote: Counting objects: 100% (140/140), done.
remote: Compressing objects: 100% (131/131), done.
remote: Total 140 (delta 67), reused 12 (delta 5), pack-reused 0 (from 0)
Receiving objects: 100% (140/140), 4.92 MiB | 14.57 MiB/s, done.
Resolving deltas: 100% (67/67), done.


In [11]:
print(os.listdir("football-qa-nlp/data"))

['results.csv', 'goalscorers.csv', '.gitkeep', 'shootouts.csv', 'former_names.csv']


In [12]:
import pandas as pd

results_df = pd.read_csv("football-qa-nlp/data/results.csv")
scorers_df = pd.read_csv("football-qa-nlp/data/goalscorers.csv")
shootouts_df = pd.read_csv("football-qa-nlp/data/shootouts.csv")
former_names_df = pd.read_csv("football-qa-nlp/data/former_names.csv")

print("✅ All datasets loaded successfully!")

✅ All datasets loaded successfully!


In [13]:
# Display the dimensions of each dataset

print("=== Dataset Dimensions ===\n")

print(f"Results Dataset: {results_df.shape}")
print(f"Goalscorers Dataset: {scorers_df.shape}")
print(f"Shootouts Dataset: {shootouts_df.shape}")
print(f"Former Names Dataset: {former_names_df.shape}")

=== Dataset Dimensions ===

Results Dataset: (49485, 9)
Goalscorers Dataset: (47855, 8)
Shootouts Dataset: (682, 5)
Former Names Dataset: (36, 4)


In [14]:
# Display dataset information

print("=== Results Dataset ===")
results_df.info()

print("\n=== Goalscorers Dataset ===")
scorers_df.info()

print("\n=== Shootouts Dataset ===")
shootouts_df.info()

print("\n=== Former Names Dataset ===")
former_names_df.info()

=== Results Dataset ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49485 entries, 0 to 49484
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        49485 non-null  object
 1   home_team   49485 non-null  object
 2   away_team   49485 non-null  object
 3   home_score  49485 non-null  int64 
 4   away_score  49485 non-null  int64 
 5   tournament  49485 non-null  object
 6   city        49485 non-null  object
 7   country     49485 non-null  object
 8   neutral     49485 non-null  bool  
dtypes: bool(1), int64(2), object(6)
memory usage: 3.1+ MB

=== Goalscorers Dataset ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47855 entries, 0 to 47854
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   date       47855 non-null  object
 1   home_team  47855 non-null  object
 2   away_team  47855 non-null  object
 3   team       47855 non-null  objec

## 3. Missing Values Analysis

In [15]:
# Check for missing values in each dataset

datasets = {
    "Results Dataset": results_df,
    "Goalscorers Dataset": scorers_df,
    "Shootouts Dataset": shootouts_df,
    "Former Names Dataset": former_names_df
}

for dataset_name, dataset in datasets.items():
    print(f"\n=== {dataset_name} ===")
    print(dataset.isnull().sum())


=== Results Dataset ===
date          0
home_team     0
away_team     0
home_score    0
away_score    0
tournament    0
city          0
country       0
neutral       0
dtype: int64

=== Goalscorers Dataset ===
date           0
home_team      0
away_team      0
team           0
scorer        48
minute       256
own_goal       0
penalty        0
dtype: int64

=== Shootouts Dataset ===
date               0
home_team          0
away_team          0
winner             0
first_shooter    422
dtype: int64

=== Former Names Dataset ===
current       0
former        0
start_date    0
end_date      0
dtype: int64


## 4. Duplicate Records Analysis

In [16]:
# Check for duplicate records in each dataset

for dataset_name, dataset in datasets.items():
    duplicate_count = dataset.duplicated().sum()
    print(f"{dataset_name}: {duplicate_count} duplicate records")

Results Dataset: 0 duplicate records
Goalscorers Dataset: 82 duplicate records
Shootouts Dataset: 0 duplicate records
Former Names Dataset: 0 duplicate records


## 5. Inspect Duplicate Records

In [17]:
# Display duplicate rows in the Goalscorers dataset

duplicate_scorers = scorers_df[scorers_df.duplicated(keep=False)]

print(f"Total duplicate rows: {len(duplicate_scorers)}")

duplicate_scorers.sort_values(
    by=["date", "home_team", "away_team", "scorer"]
).head(20)

Total duplicate rows: 128


,date,home_team,away_team,team,scorer,minute,own_goal,penalty
5105,1/16/1968,Congo,Ghana,Ghana,Osei Kofi,NaN,False,False
5106,1/16/1968,Congo,Ghana,Ghana,Osei Kofi,NaN,False,False
5128,1/19/1968,Ghana,Ivory Coast,Ivory Coast,Laurent Pokou,NaN,False,False
5129,1/19/1968,Ghana,Ivory Coast,Ivory Coast,Laurent Pokou,NaN,False,False
5124,1/19/1968,Ghana,Ivory Coast,Ghana,Wilberforce Mfum,NaN,False,False
5125,1/19/1968,Ghana,Ivory Coast,Ghana,Wilberforce Mfum,NaN,False,False
22148,1/30/2002,Burkina Faso,Ghana,Ghana,Isaac Boakye,90,False,False
22149,1/30/2002,Burkina Faso,Ghana,Ghana,Isaac Boakye,90,False,False
41775,10/12/2021,Syria,Lebanon,Lebanon,Mohamad Kdouh,45,False,False
41776,10/12/2021,Syria,Lebanon,Lebanon,Mohamad Kdouh,45,False,False


In [1]:
import pandas as pd
import numpy as np

In [2]:
!rm -rf football-qa-nlp
!git clone -b eshan-dev https://github.com/eshan14git/football-qa-nlp.git

Cloning into 'football-qa-nlp'...
remote: Enumerating objects: 140, done.
remote: Counting objects: 100% (140/140), done.
remote: Compressing objects: 100% (131/131), done.
remote: Total 140 (delta 67), reused 12 (delta 5), pack-reused 0 (from 0)
Receiving objects: 100% (140/140), 4.92 MiB | 13.78 MiB/s, done.
Resolving deltas: 100% (67/67), done.


In [3]:
results_df = pd.read_csv("football-qa-nlp/data/results.csv")
scorers_df = pd.read_csv("football-qa-nlp/data/goalscorers.csv")
shootouts_df = pd.read_csv("football-qa-nlp/data/shootouts.csv")
former_names_df = pd.read_csv("football-qa-nlp/data/former_names.csv")

In [4]:
datasets = {
    "Results Dataset": results_df,
    "Goalscorers Dataset": scorers_df,
    "Shootouts Dataset": shootouts_df,
    "Former Names Dataset": former_names_df
}

## 5. Sample Records

In [5]:
# Display the first five records of each dataset

print("=== Results Dataset ===")
display(results_df.head())

print("\n=== Goalscorers Dataset ===")
display(scorers_df.head())

print("\n=== Shootouts Dataset ===")
display(shootouts_df.head())

print("\n=== Former Names Dataset ===")
display(former_names_df.head())

=== Results Dataset ===


,date,home_team,away_team,home_score,away_score,tournament,city,country,neutral
0,1872-11-30,Scotland,England,0,0,Friendly,Glasgow,Scotland,False
1,1873-03-08,England,Scotland,4,2,Friendly,London,England,False
2,1874-03-07,Scotland,England,2,1,Friendly,Glasgow,Scotland,False
3,1875-03-06,England,Scotland,2,2,Friendly,London,England,False
4,1876-03-04,Scotland,England,3,0,Friendly,Glasgow,Scotland,False



=== Goalscorers Dataset ===


,date,home_team,away_team,team,scorer,minute,own_goal,penalty
0,7/2/1916,Chile,Uruguay,Uruguay,José Piendibene,44,False,False
1,7/2/1916,Chile,Uruguay,Uruguay,Isabelino Gradín,55,False,False
2,7/2/1916,Chile,Uruguay,Uruguay,Isabelino Gradín,70,False,False
3,7/2/1916,Chile,Uruguay,Uruguay,José Piendibene,75,False,False
4,7/6/1916,Argentina,Chile,Argentina,Alberto Ohaco,2,False,False



=== Shootouts Dataset ===


,date,home_team,away_team,winner,first_shooter
0,8/22/1967,India,Taiwan,Taiwan,NaN
1,11/14/1971,South Korea,Vietnam Republic,South Korea,NaN
2,5/7/1972,South Korea,Iraq,Iraq,NaN
3,5/17/1972,Thailand,South Korea,South Korea,NaN
4,5/19/1972,Thailand,Cambodia,Thailand,NaN



=== Former Names Dataset ===


,current,former,start_date,end_date
0,Benin,Dahomey,11/8/1959,11/30/1975
1,Burkina Faso,Upper Volta,4/14/1960,8/4/1984
2,Curaçao,Netherlands Antilles,3/3/1957,10/10/2010
3,Czechoslovakia,Bohemia,4/5/1903,1/1/1919
4,Czechoslovakia,Bohemia and Moravia,1/1/1939,5/1/1945


## 6. Descriptive Statistics

In [6]:
# Display descriptive statistics for numerical columns

print("=== Results Dataset ===")
display(results_df.describe())

print("\n=== Goalscorers Dataset ===")
display(scorers_df.describe(include="all"))

print("\n=== Shootouts Dataset ===")
display(shootouts_df.describe(include="all"))

print("\n=== Former Names Dataset ===")
display(former_names_df.describe(include="all"))

=== Results Dataset ===


,home_score,away_score
count,49485.000000,49485.000000
mean,1.757199,1.182035
std,1.773783,1.401858
min,0.000000,0.000000
25%,1.000000,0.000000
50%,1.000000,1.000000
75%,2.000000,2.000000
max,31.000000,21.000000



=== Goalscorers Dataset ===


,date,home_team,away_team,team,scorer,minute,own_goal,penalty
count,47855,47855,47855,47855,47807,47599,47855,47855
unique,4756,220,220,220,15359,133,2,2
top,10/11/2011,Brazil,Uruguay,Brazil,Cristiano Ronaldo,90,False,False
freq,145,1054,803,1079,124,2045,46929,44596



=== Shootouts Dataset ===


,date,home_team,away_team,winner,first_shooter
count,682,682,682,682,260
unique,596,188,198,183,96
top,6/3/2016,South Africa,Egypt,Egypt,Colombia
freq,5,18,17,16,11



=== Former Names Dataset ===


,current,former,start_date,end_date
count,36,36,36,36
unique,28,36,36,36
top,DR Congo,Dahomey,11/8/1959,11/30/1975
freq,4,1,1,1


## 7. Dataset Relationships

The Football Question Answering Assistant uses four related datasets.

- **Results Dataset** is the primary dataset containing international match information such as teams, scores, tournament, venue, and match date.

- **Goalscorers Dataset** provides detailed information about individual goals scored in matches. It can be linked to the Results dataset using the match date, home team, and away team.

- **Shootouts Dataset** contains information about matches decided by penalty shootouts. It also links to the Results dataset using the match date, home team, and away team.

- **Former Names Dataset** is a reference table used to map historical country names to their current names. This helps maintain consistency across datasets during preprocessing.

## 8. Exploratory Data Analysis Summary

### Key Findings

- All four datasets were successfully loaded.
- The Results dataset contains 49,485 international football matches.
- The Goalscorers dataset contains 47,855 goal records.
- The Shootouts dataset contains 682 penalty shootout records.
- The Former Names dataset contains 36 historical country name mappings.
- Missing values were identified only in the Goalscorers and Shootouts datasets.
- Duplicate records were found only in the Goalscorers dataset and will be addressed during preprocessing.
- Different date formats were identified across the datasets and will be standardized during preprocessing.
- Overall, the datasets are well structured and suitable for building the Football Question Answering Assistant.